### PR TITLE
Reduce allocations/performance overhead of draw nodes

### PR DESCRIPTION
**Pull request by smoogipoo**
1. `BufferedContainerDrawNode` now spends most of its time binding render buffers, which can be improved further, but is a larger change.
2. Edge effect draw time has been almost halved - probably also a decent representation for masking in general.

### CODE REVIEW

File: osu.Framework/Graphics/Shaders/UniformMapping.cs [hunk 1]
```diff
@@ -39,8 +39,11 @@ public void UnlinkShaderUniform(IUniform uniform)
         public void UpdateValue(ref T newValue)
         {
             Value = newValue;
-            foreach (var linked in LinkedUniforms)
-                linked.UpdateValue(this);
+
+            // Iterate by index to remove an enumerator allocation
```
Comment (peppy) [inline osu.Framework/Graphics/Shaders/UniformMapping.cs]: where are you seeing these allocs? they haven't showed up for me in dotmemory or dottrace

Comment (smoogipoo) [reply osu.Framework/Graphics/Shaders/UniformMapping.cs]: dottrace

image

(before -> after)

Comment (peppy) [reply osu.Framework/Graphics/Shaders/UniformMapping.cs]: What am i looking at there?

Comment (smoogipoo) [reply osu.Framework/Graphics/Shaders/UniformMapping.cs]: at the bottom

Comment (peppy) [reply osu.Framework/Graphics/Shaders/UniformMapping.cs]: The 4% difference?

Comment (smoogipoo) [reply osu.Framework/Graphics/Shaders/UniformMapping.cs]: That 4% difference translates into, as I said, almost 50% reduction in time spent during edge effect draw.

Comment (peppy) [reply osu.Framework/Graphics/Shaders/UniformMapping.cs]: I can't seem to find drawEdgeEffect in that profile run, am I missing osmething?

**Pull request merged**